helix-core/src/main/java/org/apache/helix/manager/zk/ParticipantManager.java

```diff
@@ -132,42 +139,102 @@ public void handleNewSession() throws Exception {
-        if (lastPos > 0) {
-          hostName = _instanceName.substring(0, lastPos);
-          port = _instanceName.substring(lastPos + 1);
+        if (!autoRegistration) {
+          LOG.info(_instanceName + " is auto-joining cluster: " + _clusterName);
+          addInstanceConfig(null);
+        } else {
+          String cloudInstanceInformationProcessorName =
+              _helixManagerProperty.getHelixCloudProperty().getCloudInfoProcessorName();
+          try {
+            // fetch cloud instance information for the instance
+            String cloudInstanceInformationProcessorClassName = CLOUD_PROCESSOR_PATH_PREFIX
+                + _helixManagerProperty.getHelixCloudProperty().getCloudProvider().toLowerCase()
+                + "." + cloudInstanceInformationProcessorName;
+            Class processorClass = Class.forName(cloudInstanceInformationProcessorClassName);
+            Constructor constructor = processorClass.getConstructor(HelixCloudProperty.class);
+            CloudInstanceInformationProcessor processor =
+                (CloudInstanceInformationProcessor) constructor
+                    .newInstance(_helixManagerProperty.getHelixCloudProperty());
+            List<String> responses = processor.fetchCloudInstanceInformation();
+
+            // parse cloud instance information for the participant
+            CloudInstanceInformation cloudInstanceInformation =
+                processor.parseCloudInstanceInformation(responses);
+            String domain = cloudInstanceInformation
+                .get(CloudInstanceInformation.CloudInstanceField.FAULT_DOMAIN.name());
+            String cloudIdInRemote = cloudInstanceInformation
+                .get(CloudInstanceInformation.CloudInstanceField.INSTANCE_SET_NAME.name());
+            String cloudIdInConfig = _configAccessor.getCloudConfig(_clusterName).getCloudID();
+
+            // validate that the instance is auto registering to the correct cluster
+            if (!cloudIdInRemote.equals(cloudIdInConfig)) {
+              throw new IllegalArgumentException(String.format(
+                  "cloudId in config: %s is not consistent with cloudId from remote: %s. The instance is auto registering to a wrong cluster.",
+                  cloudIdInConfig, cloudIdInRemote));
+            }
+            addInstanceConfig(domain);
+          } catch (ClassNotFoundException ex) {
+            throw new HelixException(
+                "Passed cloud instance information processor class is not found: "
+                    + cloudInstanceInformationProcessorName, ex);
+          } catch (NoSuchMethodException ex) {
+            throw new HelixException("Failed to get the constructor for the class: "
+                + cloudInstanceInformationProcessorName, ex);
+          } catch (InstantiationException | IllegalAccessException | InvocationTargetException ex) {
+            throw new HelixException("Failed to create a new instance for the class: "
```

Comment:
Any reason to differentiate these exceptions? They looked like instantiation failure.